src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +309,245 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+        name,
+        method,
+        range,
+    });
+}
+
+
+void PrometheusSerializer::Annotate(
+        std::ostream& out, std::string& metric_name, std::string&& metric_type) const {
+    out << "# HELP " << metric_name << " Record all " << metric_type << " about nebula" << "\n";
+    out << "# TYPE " << metric_name << " " << metric_type << "\n";
+}
+
+// Write a double as a string, with proper formatting for infinity and NaN
+void PrometheusSerializer::WriteValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::WriteValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::WriteValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::WriteHead(std::ostream& out, const std::string& family,
+                const std::vector<Label>& labels, const std::string& suffix,
+                const std::string& extraLabelName,
+                const T extraLabelValue) const {
+    out << family << suffix;
+    if (!labels.empty() || !extraLabelName.empty()) {
+        out << "{";
+        const char* prefix = "";
+
+        for (auto& lp : labels) {
+            out << prefix << lp.name << "=\"";
+            WriteValue(out, lp.value);
+            out << "\"";
+            prefix = ",";
+        }
+        if (!extraLabelName.empty()) {
+            out << prefix << extraLabelName << "=\"";
+            WriteValue(out, extraLabelValue);
+            out << "\"";
+        }
+        out << "}";
+    }
+    out << " ";
+}
+
+// Write a line trailer: timestamp
+void PrometheusSerializer::WriteTail(std::ostream& out, const std::int64_t timestamp_ms) const {
+    if (timestamp_ms != 0) {
+        out << " " << timestamp_ms;
+    }
+    out << "\n";
+}
+
+// TODO(shylock) transform nebula stats to Prometheus::ClientMetric/MetricFamily
+// using the serrializing from 3rd-SDK
+void StatsManager::PrometheusSerialize(std::ostream& out) /*const*/ {
+    std::locale saved_locale = out.getloc();
+    out.imbue(std::locale::classic());
+    for (auto& index : nameMap_) {
+        auto parsedName = parseMetricName(index.first);
```

Comment:
Maybe we can check style when we register, so that we don't need check it here?

Reply:
Not the check, I add this to determine the metric information such as `method` or `time range`,it's optional name rule.

Reply:
I mean many place check `parsedName.ok()`, and then have different process logic. If we already check the metrics when we register, how can we fail at `parseMetricName`.

Reply:
BTW, it seems you can have a map store all parsedName, just collect them when register metrics, no need to parse each time?

Reply:
Good idea.